homeassistant/components/nws/weather.py

```diff
@@ -0,0 +1,378 @@
+    longitude = config.get(CONF_LONGITUDE, hass.config.longitude)
+    station = config.get(CONF_STATION)
+    api_key = config[CONF_API_KEY]
+    mode = config[CONF_MODE]
+
+    if None in (latitude, longitude):
+        _LOGGER.error("Latitude/longitude not set in Home Assistant config")
+        return ConfigEntryNotReady
+
+    websession = async_get_clientsession(hass)
+    # ID request as being from HA, pynws prepends the api_key in addition
+    api_key_ha = f"{api_key} homeassistant"
+    nws = SimpleNWS(latitude, longitude, api_key_ha, mode, websession)
+
+    _LOGGER.debug("Setting up station: %s", station)
+    try:
+        await nws.set_station(station)
+    except ERRORS as status:
+        _LOGGER.error(
+            "Error getting station list for %s: %s", (latitude, longitude), status
+        )
+        raise PlatformNotReady
+
+    _LOGGER.debug("Station list: %s", nws.stations)
+    _LOGGER.debug(
+        "Initialized for coordinates %s, %s -> station %s",
+        latitude,
+        longitude,
+        nws.station,
+    )
+
+    async_add_entities([NWSWeather(nws, mode, hass.config.units, config)], True)
+
+
+class NWSWeather(WeatherEntity):
+    """Representation of a weather condition."""
+
+    def __init__(self, nws, mode, units, config):
+        """Initialise the platform with a data instance and station name."""
+        self.nws = nws
+        self.station_name = config.get(CONF_NAME, self.nws.station)
+        self.is_metric = units.is_metric
+        self.mode = mode
+
+        self.observation = None
+        self._forecast = None
+
+    @Throttle(MIN_TIME_BETWEEN_UPDATES)
+    async def async_update(self):
+        """Update Condition."""
+        _LOGGER.debug("Updating station observations %s", self.nws.station)
+        try:
+            await self.nws.update_observation()
+        except ERRORS as status:
+            _LOGGER.error(
+                "Error updating observation from station %s: %s",
+                self.nws.station,
+                status,
+            )
+        else:
+            self.observation = self.nws.observation
+        _LOGGER.debug("Updating forecast")
+        try:
+            await self.nws.update_forecast()
+        except ERRORS as status:
+            _LOGGER.error(
+                "Error updating forecast from station %s: %s", self.nws.station, status
+            )
+        else:
+            self._forecast = self.nws.forecast
+        return
+
+    @property
+    def attribution(self):
+        """Return the attribution."""
+        return ATTRIBUTION
+
+    @property
+    def name(self):
+        """Return the name of the station."""
+        return self.station_name
+
+    @property
+    def temperature(self):
+        """Return the current temperature."""
+        temp_c = None
+        if self.observation:
+            temp_c = self.observation.get("temperature")
+        if temp_c:
+            return convert_temperature(temp_c, TEMP_CELSIUS, TEMP_FAHRENHEIT)
```

Comment:
Why not report the temperature in celsius and let the core convert as needed?

Reply:
The NWS API does not allow for a choosable unit system.  NWS provides the observation in C and the forecast in F.  Wonderful isn't it? Given the existence of unit conversion in HA, it is cleanest to do here.

Reply:
I could move the conversion to the forecast if you'd prefer.

Reply:
The temperature will be converted as needed here:
https://github.com/home-assistant/home-assistant/blob/f30e54f01e0460f08ff3d470ca45612eb9d860d8/homeassistant/components/weather/__init__.py#L126-L127

Reply:
Ok, now I get it. Either we convert here or in forecast. Let's just keep it.